pkg/server/apiextensions.go

```diff
@@ -33,135 +33,201 @@ import (
-			// on its CRs with kubectl (because it relies on discovery, and returning [] when we can't
-			// find the ClusterWorkspace would mean CRDs from this logical cluster wouldn't be in discovery).
-			return nil, err
+	for _, apiBinding := range apiBindings {
+		if apiBinding.ClusterName != cluster.Name {
+			continue
+		}
+		if apiBinding.Status.Phase != apisv1alpha1.APIBindingPhaseBound {
+			// TODO(ncdc): what if it's in the middle of rebinding and it takes some time - we won't include any
+			// CRDs that were previously bound...
+			continue
 		}
 
-		if workspace != nil && workspace.Spec.InheritFrom != "" {
-			if workspace.Spec.InheritFrom == helper.RootCluster {
-				// HACK: allow inheriting from the OrganizationCluster logical cluster
-				inheriting = true
-				inheritFrom = helper.RootCluster
-			} else {
-				// Make sure the source workspace exists
-				wsName := workspace.Spec.InheritFrom
-				orgClusterName := orgName
-				if strings.ContainsRune(workspace.Spec.InheritFrom, ':') {
-					orgClusterName, err = helper.ParentClusterName(workspace.Spec.InheritFrom)
-					if err != nil {
-						return nil, err
-					}
-					orgName, wsName, err = helper.ParseLogicalClusterName(workspace.Spec.InheritFrom)
-					if err != nil {
-						return nil, err
-					}
-				}
-				sourceWorkspaceKey := helper.WorkspaceKey(orgClusterName, wsName)
-				_, err := c.workspaceLister.Get(sourceWorkspaceKey)
-				switch {
-				case err == nil:
-					inheriting = true
-					inheritFrom = helper.EncodeOrganizationAndClusterWorkspace(orgName, wsName)
-				case apierrors.IsNotFound(err):
-					// A NotFound error is ok. It means we can't inherit but we should still proceed below to list.
-				default:
-					// Only error if there was a problem checking for workspace existence
-					return nil, err
-				}
+		for _, boundResource := range apiBinding.Status.BoundResources {
+			crdKey := clusters.ToClusterAwareKey("system:bound-crds", boundResource.Schema.UID)
+			crd, err := c.crdLister.GetWithContext(ctx, crdKey)
+			if err != nil {
+				klog.Errorf("Error getting bound CRD %q: %v", crdKey, err)
+				continue
+			}
+
+			if seen.Has(crdName(crd)) {
+				// Came from system
+				klog.Infof("Skipping APIBinding CRD %s|%s because it came in via system CRDs", crd.ClusterName, crd.Name)
+				continue
 			}
+
+			ret = append(ret, crd)
+			seen.Insert(crdName(crd))
 		}
 	}
 
-	var ret []*apiextensionsv1.CustomResourceDefinition
 	crds, err := c.crdLister.List(labels.Everything())
 	if err != nil {
 		return nil, err
 	}
 	for i := range crds {
 		crd := crds[i]
-		if crd.ClusterName == cluster.Name || (inheriting && crd.ClusterName == inheritFrom) {
-			ret = append(ret, crd)
+		if crd.ClusterName != cluster.Name {
+			continue
+		}
+		if seen.Has(crdName(crd)) {
+			klog.Infof("Skipping local CRD %s|%s because it came in via APIBindings or system CRDs", crd.ClusterName, crd.Name)
+			continue
 		}
```

Comment:
and here: system CRDs trump over bindings trumping over CRDs ?

Maybe worth a comment on both blocks